### PR TITLE
feat(data-warehouse): linking views to posthog models

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

- as a user, I would like to be able to attach views to existing posthog models such that I can query the posthog model as if it had the view's fields (similar to how you can query person_distinct_id fields directly from events where the joins are handled implicitly)

## Changes

- add backend model and api for creating a link between a view and existing posthog model
- fields will be queryable in hogql

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/warehouse/models/view_link.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from posthog.models.utils import UUIDModel, CreatedMetaFields, DeletedMetaFields
+from django.db import models
+from posthog.models.team import Team
+from .datawarehouse_saved_query import DataWarehouseSavedQuery
+
+
+class DataWarehouseViewLink(CreatedMetaFields, UUIDModel, DeletedMetaFields):
```
Comment (EDsCODE) [inline posthog/warehouse/models/view_link.py:7]: Data model seems a bit flimsy as it's a bunch of string matching (as opposed to having an Enum of possible tables/fields)

File: posthog/hogql/database/database.py [hunk 2]
```diff
@@ -115,6 +125,29 @@ def create_hogql_database(team_id: int) -> Database:
...
+        join_expr = ast.JoinExpr(table=parse_select(view.saved_query.query["query"]))
+        join_expr.join_type = "INNER JOIN"
+        join_expr.alias = to_table
+        join_expr.constraint = ast.JoinConstraint(
+            expr=ast.CompareOperation(
+                op=ast.CompareOperationOp.Eq,
+                left=ast.Field(chain=[from_table, view.from_join_key]),
+                right=ast.Field(chain=[to_table, view.to_join_key]),
+            )
+        )
```
Comment (Twixes) [inline posthog/hogql/database/database.py:145]: Nit: These can be just provided to the `ast.JoinExpr()` constructor, same as `table`

File: posthog/hogql/database/database.py [hunk 3]
```diff
@@ -115,6 +125,29 @@ def create_hogql_database(team_id: int) -> Database:
     return database
 
 
+def determine_join_function(view):
```
Comment (Twixes) [inline posthog/hogql/database/database.py:128]: Given this is so tightly coupled with the `view`, might be better as a
```python
@property
def join_function
```
on `DataWarehouseViewLink`?

**Review by Twixes — CHANGES_REQUESTED** [review]
A few small rough edges, and a safety concern

File: posthog/warehouse/query.py
[the remark this answers is not in the document]
Comment (EDsCODE) [reply posthog/warehouse/query.py:1]: No idea, it was there for a while for no reason so I'm deleting it

File: posthog/hogql/database/models.py [hunk 4]
```diff
@@ -149,3 +148,9 @@ class SavedQuery(Table):
 
     query: str
     name: str
+
+    def to_printed_clickhouse(self, context):
+        return self.name
+
+    def to_printed_hogql(self):
+        return self.name
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply posthog/hogql/database/models.py]: done

File: posthog/hogql/database/database.py [hunk 5]
```diff
@@ -103,6 +103,16 @@ def create_hogql_database(team_id: int) -> Database:
         database.events.fields["person"] = FieldTraverser(chain=["poe"])
         database.events.fields["person_id"] = StringDatabaseField(name="person_id")
 
+    for view in DataWarehouseViewLink.objects.filter(team_id=team.pk).exclude(deleted=True):
+        table = database.get_table(view.table)
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply posthog/hogql/database/database.py:107]: Added logic/update data model to be able to cascade deletion of s3table to all related views. Will account for this in UI so it doesn't seem buggy

File: posthog/warehouse/models/datawarehouse_saved_query.py [hunk 6]
```diff
@@ -45,6 +48,24 @@ def get_columns(self) -> Dict[str, str]:
         types = response.get("types", {})
         return dict(types)
 
+    @property
+    def s3_tables(self):
+        from posthog.hogql.parser import parse_select
+        from posthog.hogql.context import HogQLContext
+        from posthog.hogql.database.database import create_hogql_database
+        from posthog.hogql.resolver import resolve_types
+        from posthog.models.property.util import S3TableVisitor
```
Comment (Twixes) [inline posthog/warehouse/models/datawarehouse_saved_query.py:57]: Do _all_ of these need to be imported here? A bit suspicious

**Review by Twixes — APPROVED** [review]
LGTM!

**Review by mariusandra — APPROVED** [review]
Generally looking good! Some comments below, but non blocking:

- If I just see a "save as view" button, I don't know what that does. Perhaps a "(i)" next to it with a tooltip explaining what's what?
- The error discards the unsaved state:

- We could add an alias automatically for simple field/property access

**Comment by EDsCODE** [comment]
@mariusandra Added comments to https://github.com/PostHog/posthog/pull/16993 to address UI issues together

**Comment by EDsCODE** [comment]
Will let this test run again but should be irrelevant. Looks like a cache failure. Also, this PR has near 0 impact on UI

**Pull request merged**